pkg/dataobj/metastore/toc_writer.go

```diff
@@ -96,87 +97,104 @@ func (m *TableOfContentsWriter) WriteEntry(ctx context.Context, dataobjPath stri
 		return err
 	}
 
+	var globalMinTime, globalMaxTime time.Time
+	for _, timeRange := range tenantTimeRanges {
+		if globalMinTime.IsZero() || timeRange.MinTime.Before(globalMinTime) {
+			globalMinTime = timeRange.MinTime
+		}
+		if globalMaxTime.IsZero() || timeRange.MaxTime.After(globalMaxTime) {
+			globalMaxTime = timeRange.MaxTime
+		}
+	}
+
 	// Work our way through the metastore objects window by window, updating & creating them as needed.
 	// Each one handles its own retries in order to keep making progress in the event of a failure.
-	prefix := storagePrefixFor(m.cfg.Storage, m.tenantID)
-	for metastorePath := range iterTableOfContentsPaths(m.tenantID, minTimestamp, maxTimestamp, prefix) {
-		m.backoff.Reset()
-		for m.backoff.Ongoing() {
-			err = m.bucket.GetAndReplace(ctx, metastorePath, func(existing io.ReadCloser) (io.ReadCloser, error) {
-				if existing != nil {
-					defer existing.Close()
-				}
+	for _, timeRange := range tenantTimeRanges {
+		prefix := storagePrefixFor(m.cfg.Storage, string(timeRange.Tenant))
+		for tocPath, tocTimeRange := range iterTableOfContentsPaths(string(timeRange.Tenant), globalMinTime, globalMaxTime, prefix) {
+			m.backoff.Reset()
```

Comment:
Nit for future change: as we don't use the backoff between calls, I feel that we should allocate a new backoff on the stack for each iteration, instead of resetting a stateful backoff that lives for the duration of the writer.

There should be nothing here that escapes to the heap, and so it should be very fast to stack allocate lots of these with no GC overhead.

I can benchmark it later to confirm if my hypothesis is correct.

Reply:
I'm not against that. I'll fix this in the next PR though, I think!